docs/import_specifications.ADR.md

```diff
@@ -0,0 +1,186 @@
+extremely unlikely that there will be a large enough set of data types that URL length limits
+will be reached so this seems adequate. The StS will respond with the contents of the files
+parsed into a structure that is similar to that of the input spec, mapped by data type:
+
+```
+Headers:
+Authorization: <KBase token>
+
+GET /bulk_specification/?files=<file1.csv[,file2.csv,...]>
+
+Response:
+{"types": {
+    "gff_metagenome": [
+        {"fasta_file": "mygenome.fa",
+         "gff_file": "mygenome.gff3",
+         "genome_name": "mygenomeobject",
+         "source": null,
+         "release": null,
+         "genetic_code: 11,
+         "generate_missing_genes: 0
+         },
+        {"fasta_file": "mygenome2.fa",
+         "gff_file": "mygenome2.gff3",
+         "genome_name": "mygenomeobject2",
+         "source": "yermumspoo",
+         "release": 30456,
+         "genetic_code: 11,
+         "generate_missing_genes: 1
+         },
+         ...
+    ],
+    "sra_reads": [
+        ...
+    ]
+    ...
+}
+```
+
+The order of the input structures MUST be the same as the order in the input files.
+
+Notably, the service will provide the contents of the files as is and will not perform most error
+checking, including for missing or unknown input parameters. Most error checking will be performed
+in the bulk import cell configuration tab like other uploads, allowing for a consistent user
+experience.
+
+### Missing files
+
+If the StS cannot find one or more files it will return a 404 with the body contents:
+```
+{"error": "cannot_find_all_files",
+ "missing_files": [<list of missing filenames>]
+}
+```
+
+### Unparseable files
+
+If the StS cannot parse one or more of the files it will return a 400 with the body contents:
+```
+{"error": "cannot_parse_all_files",
+ "unparseable_files": [{"filename": <filename>,
+                        "reason": <message>
+                        },
+                        ...
+                       ]
+}
+```
+The service will check that the data type is valid and that rows >=2 all have the same number of
+entries, but will not do further error checking.
+
+Note in this case the service MUST log the stack trace along with the filename for each invalid
+file.
+
+### Other errors
+
+If any other error occurs, a general 500 error will be reported:
+```
+{"error": "unexpected_error",
+ "message": <error message here>
+}
+```
+
+## Alternatives explored
+
+* We considered parsing the files in the Narrative backend (pulling the files as is from the StS)
+  but the current design allows other services and UIs reusing the parsing code. There doesn't
+  otherwise seem to be a strong reason to include the code one place or the other so we decided
+  to include it in the StS.
+
+## Questions
+
+* Should it be an error to submit multiple files or tabs for the same type? If not,
+  how should the diffent files be denoted and ordered?
```

Comment:
One use case for multiple files or tabs that I can think of is if users organized their data that way for separate datasets. But they could always upload each dataset separately

Reply:
or C&P into a single file

Reply:
The only case I can think of where they would want this is if the have a multiple-parameter upload (e.g. 5 Illumina reads files and 5 PacBio reads files) and they wanted to do that simultaneously, which wouldn't be possible under this implementation anyway, would it? To me the best options would be:
1. If possible, merge the files. If that would take basically any work on our end, then skip and go to 2.
2. Give a message like "Multiple Import Specification files selected for [type]. Please remove excess files or merge into a single file."

Reply:
You could do it if both sets of reads 1) are interleaved or non-interleaved and 2) have the same optional parameters, if any.

> If possible, merge the files.

It's possible, but thinking ahead, how would that look in the spreadsheet view in the front end that's on the roadmap (but not in the xCV MAP)? That's what has me concerned about this, otherwise I would've just merged the files.

Reply:
It also makes user debugging harder as they have to figure out which row in the UI goes to which file

Reply:
Maybe we say 1 file for 1 type in the MVP and consider multliple files later, if at all

Reply:
1 file 1 type seems good to me for the MVP and even a bit beyond. I think multiple files per type is going to be an edge case. They way I viewed merged them was concatenating one to the bottom of the other, but I didn't take into account displaying the errors. I say skip it and make the user copy-paste their own files.

Reply:
Added 1 file/tab per type as an answer for this question